### PR TITLE
feat: added login 

### PR DESCRIPTION
**Pull request by Vsevolod1020**

### CODE REVIEW

**Review by Ars6300 — APPROVED** [review]

File: src/app/core/authentication/authentication.service.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  constructor(
+    private http: HttpClient,
+    private router: Router
+  ){}
+
+  login(user: { email: any; password: any; }){
+    // return this.http.post(`https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${environment.apiKey}`, user)
+    return this.http.post(environment.api_URL, user)
+    .pipe(
+      tap(this.controller)
```
Comment (nattallius) [inline src/app/core/authentication/authentication.service.ts]: Maybe it's better to call this method setCookies?

File: src/app/core/authentication/authentication.service.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  logout(){
+    document.cookie = `test=test; max-age=-1`;
+  }
+
+  get token(){
+    const cookieDate = document.cookie
+    if (!!cookieDate) {
+      return cookieDate.split('=')[1]
+    }
```
Comment (nattallius) [inline src/app/core/authentication/authentication.service.ts]: document.cookie returns **ALL** cookies for this domain as a single string. To check and return cookie with certain name you should parse this string. Something like that 
`document.cookie.split(';').find((cookie) => cookie.startsWith(tokenName));`

File: src/app/core/authentication/authentication.service.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    return this.logout()
```
Comment (nattallius) [inline src/app/core/authentication/authentication.service.ts]: If you create a getter called get token, then it should return only associated data and doesn't perform any other side actions, like logout.

File: src/app/core/authentication/authentication.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  private controller(response: any){
+    if (response) {
+      // document.cookie = `token=${response.Idtoken}; max-age=3600`;
+      document.cookie = `token=${response.token}; max-age=3600`;
+    }
+  }
+
+  logout(){
```
Comment (nattallius) [inline src/app/core/authentication/authentication.service.ts]: It's better to rename it to deleteCookie

File: src/app/core/guard/guard.guard.ts [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
     route: ActivatedRouteSnapshot,
     state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
-    return true;
+      if (this.auth.isAuthenticated()) {
+        return true;
+      } else {
+        this.auth.logout()
+        this.router.navigate(['/']) 
+      }
+      return true
```
Comment (nattallius) [inline src/app/core/guard/guard.guard.ts]: return false?

File: src/app/core/guard/guard.guard.ts [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
+  ){}
   canActivate(
     route: ActivatedRouteSnapshot,
     state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
-    return true;
+      if (this.auth.isAuthenticated()) {
+        return true;
+      } else {
+        this.auth.logout()
+        this.router.navigate(['/'])
```
Comment (nattallius) [inline src/app/core/guard/guard.guard.ts]: Add return false  after navigation

File: src/app/modules/grammar/grammar-routing.module.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import { NgModule } from '@angular/core';
 import { RouterModule, Routes } from '@angular/router';
 import { GRAMMAR_PATH, QUESTION_GRAMMAR_PATH } from 'src/app/app-routing.constants';
+import { GuardGuard } from 'src/app/core/guard/guard.guard';
```
Comment (nattallius) [inline src/app/modules/grammar/grammar-routing.module.ts]: Please, rename it to logical name, e.g. _AuthGuard_

File: src/app/pages/login/components/login/login.component.ts [hunk 8]
```diff
@@ -1,27 +1,55 @@
...
-    console.log('Form Valid:' + userForm.valid);
-    console.log('Form Submitted:' + userForm.submitted);
+    const user = {
+      email: this.form.value.email,
+      password: this.form.value.password,
+    }
+  
+    this.auth.login(user).subscribe( res => {
+      // console.log(res)
+      this.form.reset
```
Comment (nattallius) [inline src/app/pages/login/components/login/login.component.ts]: this.form.reset()?

**Review by nattallius — COMMENTED** [review]
👍

File: src/app/core/authentication/authentication.service.ts (hunk 1 above)
Comment (Vsevolod1020) [reply src/app/core/authentication/authentication.service.ts]: ok I will rename this

File: src/app/core/authentication/authentication.service.ts (hunk 4 above)
Comment (Vsevolod1020) [reply src/app/core/authentication/authentication.service.ts]: ok

File: src/app/core/authentication/authentication.service.ts (hunk 2 above)
Comment (Vsevolod1020) [reply src/app/core/authentication/authentication.service.ts]: But if i'm not mistaken this document.cookie.split(';').find((cookie) => cookie.startsWith(tokenName)); will return name and value? I guess I only need to return the value, don't I?

File: src/app/core/authentication/authentication.service.ts (hunk 3 above)
Comment (Vsevolod1020) [reply src/app/core/authentication/authentication.service.ts]: No problem, I can remove logout. But there was a little checking in "get token". If the time of cookies has expired, then just clear the cookies. And if there is something, then return the token. It turns out then just cookies will not be cleared without logout

File: src/app/core/guard/guard.guard.ts (hunk 6 above)
Comment (Vsevolod1020) [reply src/app/core/guard/guard.guard.ts]: ok

File: src/app/core/guard/guard.guard.ts (hunk 5 above)
Comment (Vsevolod1020) [reply src/app/core/guard/guard.guard.ts]: ok)

File: src/app/modules/grammar/grammar-routing.module.ts (hunk 7 above)
Comment (Vsevolod1020) [reply src/app/modules/grammar/grammar-routing.module.ts]: Okay, that was the default name. But I'll rename it. Thank you. I would also like to ask, I don't have a guard working on Speaking, Writing and Listening routing modules. But I specified them as canActivate: [GuardGuard].

File: src/app/pages/login/components/login/login.component.ts (hunk 8 above)
Comment (Vsevolod1020) [reply src/app/pages/login/components/login/login.component.ts]: Oh, I forgot to put brackets, thank you

**Review by DobrovskayaElena — APPROVED** [review]

**Pull request merged**